### DISCUSSION TITLE
Getting weather data into DB

### DISCUSSION THREAD

**Discussion by CaiusX** [mcguirepr89/BirdNET-Pi#78] [Q&A] [answered]
Hi @mcguirepr89 

As discussed yesterday, I've started pulling in the temperature into the BirdNET-Pi BirdsDB.txt file through the analyze.py file.  It gets the temperature based on lat/long from wttr.in whenever there is a positive detection and writes it to the BirdsDB.txt

- the specific code is from line 241

```
#Get temperature from internet based on latitude & longitude                       
                        try:
                            temperature = requests.get('https://wttr.in/args.lat,args.lon?format=3').text.replace(" ",";").split(';')[3][:-3][1:]
                        except:
                            temperature = "N/A"
                        
                        rfile.write(str(current_date) + ';' + str(current_time) + ';' + entry[0].replace('_', ';') + ';' \
                        + str(entry[1]) +";" + str(args.lat) + ';' + str(args.lon) + ';' + str(min_conf) + ';' + str(week) + ';' \
                        + str(sensitivity) +';' + str(temperature) + '\n')
```

Also "import requests" is needed

The try/except is so that the system doesn't fall over when losing internet - tested and seems to work.

I have substituted the "Overlap" field with the Temperature so as not to mess with the structure, but will probably be better to add as a new field.

For some reason this isn't pulling into the SQL database - maybe you can have a look?
.

Best

**Comment by mcguirepr89** [comment, marked as answer]
@CaiusX This is awesome -- I really appreciate you getting this started.

> I have substituted the "Overlap" field with the Temperature so as not to mess with the structure, but will probably be better to add as a new field.
For some reason this isn't pulling into the SQL database - maybe you can have a look?

The reason the DB is not accepting those values is due to its expected data type structure:
```
+------------+--------------+------+-----+---------+-------+
| Field      | Type         | Null | Key | Default | Extra |
+------------+--------------+------+-----+---------+-------+
| Date       | date         | YES  |     | NULL    |       |
| Time       | time         | YES  |     | NULL    |       |
| Sci_Name   | varchar(100) | NO   |     | NULL    |       |
| Com_Name   | varchar(100) | NO   |     | NULL    |       |
| Confidence | float        | YES  |     | NULL    |       |
| Lat        | float        | YES  |     | NULL    |       |
| Lon        | float        | YES  |     | NULL    |       |
| Cutoff     | float        | YES  |     | NULL    |       |
| Week       | int(11)      | YES  |     | NULL    |       |
| Sens       | float        | YES  |     | NULL    |       |
| Overlap    | float        | YES  |     | NULL    |       |
+------------+--------------+------+-----+---------+-------+
```
The "Overlap" field is expecting a "Float" value. Since `wttr.in` spits out a string with numbers and letters and symbols as its results, creating a new field with a data type of "varchar(50)" should work. 

To add the new column to the table, do the following:

1. Issue `sudo mysql birds` in the terminal.
2. Once at the SQL command line, enter `ALTER TABLE detections ADD Weather varchar(50) after Overlap;`
3. Then `exit`.

I'm unsure whether the weather icons are going to cause issues or if the degree symbol will. See my results below for what I mean:
Call: `curl wttr.in/35.0,-80\?format=3`
Results:
```
35.0,-80: ⛅️  +71°F
```

One option may be to convert the weather symbol to plain text in the wttr.in results (if the symbols cause problems):
Call: `curl wttr.in/35.0,-80\?format="%C+%t"`
Results:
```
Partly cloudy +71°F
```

I will likely begin tinkering with this early next week as well, so let me know if the above helps and we'll go from there!

**Comment by mcguirepr89** [comment]
Something occurred to me as I was drifting off to sleep last night -- wttr.in has request limits that will definitely be hit in systems like yours that have so many detections each day.

With this in mind, it is probably going to be required that there be a separate 'weather' table that can be used to cross reference the 'detections' table. This will allow us to take weather measurements at regular intervals through the day (every hour, half hour, whatever doesn't hit the request limit) and have an overview of the weather data over time instead of making a request each time detections are written to the database. This will also allow us to use allllll of the information that wttr.in can supply, including lunar calendar data, barometric pressure, sunrise/sunset/dawn/dusk/zenith, and everything below:
_I don't know everything that affects birds, but more data will have answers_
```
    c    Weather condition,
    C    Weather condition textual name,
    x    Weather condition, plain-text symbol,
    h    Humidity,
    t    Temperature (Actual),
    f    Temperature (Feels Like),
    w    Wind,
    l    Location,
    m    Moon phase 🌑🌒🌓🌔🌕🌖🌗🌘,
    M    Moon day,
    p    Precipitation (mm/3 hours),
    P    Pressure (hPa),

    D    Dawn*,
    S    Sunrise*,
    z    Zenith*,
    s    Sunset*,
    d    Dusk*,
    T    Current time*,
    Z    Local timezone.

(*times are shown in the local timezone)
```
For displaying that data, we'll just be altering some of the SQL statements to `SELECT Temp, Humidity, Zenith, Whatever FROM weather WHERE Date == "12/25/2021"`

I'm still thinking on this, because it will make sorting through data _by weather type_ less straightforward than it would be if the info were alongside each detection.

:thinking:

**Reply by jmtuts** [reply]
Did this get tabled?  I don't see a weather table in the Birds.db database
